### ISSUE TITLE
jax.numpy.sort performance regression

### ISSUE THREAD

**Issue by drewjenks01** [jax-ml/jax#30296]
### Description

Hello, I have recently upgraded my GPU training code to use jax 0.5.3 from 0.4.25. I have some sorting logic within my code which uses the jax numpy sort:
```
result = jnp.linalg.norm(delt, ord=2, axis=-1)
result = jnp.sort(result, axis=1)
```
I noticed that after I upgraded and profiled this code, the sort() function takes 3x slower. I tried using jax.lax.sort() and that was even slower. Is this a known bug? Would this be fixed by upgrading Jax to a more recent version?

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.5.3
jaxlib: 0.5.3
numpy:  1.26.4
python: 3.10.16 (zoox-r0@redacted) [GCC 11.4.0]
device info: GRID T4-8Q-1, 1 local devices"
process_count: 1
platform: uname_result(system='Linux', node='svdi-u8q-547', release='5.15.0-143-generic', version='#153~20.04.1-Ubuntu SMP Fri Jun 20 11:54:24 UTC 2025', machine='x86_64')


$ nvidia-smi
Fri Jul 18 01:07:52 2025       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.183.01             Driver Version: 535.183.01   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  GRID T4-8Q                     On  | 00000000:02:00.0 Off |                    0 |
| N/A   N/A    P0              N/A /  N/A |    493MiB /  8192MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A      2986      G   /usr/lib/xorg/Xorg                          118MiB |
|    0   N/A  N/A      3253      G   /usr/bin/gnome-shell                        267MiB |
|    0   N/A  N/A   1815837      C   ...ace_test.runfiles/python3.10/python      107MiB |
+---------------------------------------------------------------------------------------+
```

**Comment by drewjenks01** [comment]
By the way, this is after already calling jax.jit

**Comment by jakevdp** [comment]
Can you say more about how you're measuring the timings of these microbenchmarks? Are you accounting for the issues discussed at https://docs.jax.dev/en/latest/faq.html#benchmarking-jax-code? (i.e. asynchronous dispatch, data placement, compilation cost, etc.)

A [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) would be helpful here.

**Comment by drewjenks01** [comment]
Thanks for the response @jakevdp!

We use xprof to profile the code -- here are screenshots of profiling the sort() method in 0.4.25 and 0.5.3:

<img width="2108" height="1214" alt="Image" src="https://github.com/user-attachments/assets/d90e00b9-a01c-4ce2-8fcc-60f2d3c7e05c" />
<img width="2648" height="1272" alt="Image" src="https://github.com/user-attachments/assets/b600d1bf-b800-4054-a7f3-ef9988773614" />

**Comment by jakevdp** [comment]
Thanks - offhand I don't know of any changes to sort behavior between 0.4.25 and 0.5.3, but those versions are both quite old now (Feb 2024 to Mar 2025).

Have you tried with more recent JAX versions?

**Comment by drewjenks01** [comment]
We are unable to try Jax 0.6.2 since we have cudnn 9.4 and it seems like it requires 9.8 (we tried with jax[cuda12] but this conflicted with torch 2.6 so we had to use jax[cuda12_local] but this ended up failing cuda setup since we have cudnn 9.4 and it was expecting 9.8).

Do you have any intuition into what could cause this or what would be good to do to debug? Something with generated XLA kernels?

**Comment by jakevdp** [comment]
On the JAX side, this looks like the only relevant change to sort lowering in that period of time: https://github.com/jax-ml/jax/pull/22211. It's not clear to me if that might affect performance.

**Comment by drewjenks01** [comment]
We did some experimenting on isolated sort() call with different input shapes between Jax 0.4.25, 0.5.0, and 0.5.3. We found that axis 1 sorting is significantly slower in 0.5.0 and 0.5.3 ![image](https://github.com/user-attachments/assets/1c0e3ef1-75a0-4acc-b914-08989457a979)![image](https://github.com/user-attachments/assets/2b9d614d-de3c-4d7e-b972-25fb6aa19fdd)![image](https://github.com/user-attachments/assets/8a8d3bfc-fe3a-4eef-a7dd-620efed6cf8d)

**Comment by drewjenks01** [comment]
Interestingly, transposing the input and then sorting on axis 0 regains a lot of the regression
